File: src/logic.rs

pub const MAX_SLICES : usize = 32;

#[derive(Copy,Clone)]
pub struct SliceStack {
    count : i32,
    color_count : i32,
    slice_type : [u8; MAX_SLICES]
}

/*
impl Default for SliceStack {
    fn default() -> SliceStack {
        SliceStack { count:0, color_count:0, slice_type:[0; MAX_SLICES] }
    }
}
*/

/* START PRNG HELPERS */
pub struct PrngCtxGaloisLsfw {
    value : u32
}
pub fn prng_seed_galois_lsfw(ctx : &mut PrngCtxGaloisLsfw, s : u32) {
    ctx.value = s;
}

pub fn prng_galois_lsfw(ctx : &mut PrngCtxGaloisLsfw) -> u32 {
    let shifted = ctx.value >> 1;
    let negated = (-((ctx.value & 1) as i32)) as u32;
    let anded = negated & 0x80200003;
    ctx.value = shifted ^ anded;
    return ctx.value;
}

pub fn prng_galois_lsfw_int_minmax(ctx : &mut PrngCtxGaloisLsfw, min : i32,
                                   max : i32) -> i32 {
    use std::u32;
    const INV_INT_MAX : f32 = 1.0 / ((u32::MAX) as f32);
    let f : f32 = (prng_galois_lsfw(ctx) as f32) * INV_INT_MAX;
    let frange : f32 = (max-min) as f32;
    return ((f * frange + 0.5) as i32) + min;
}

static mut l_ctx : PrngCtxGaloisLsfw = PrngCtxGaloisLsfw {value:34};
fn fake_rand(min : i32, max : i32) -> i32 {
    unsafe {
        return prng_galois_lsfw_int_minmax(&mut l_ctx, min, max);
    }
}
/* END PRNG HELPERS */

pub fn ss_init(ss : &mut SliceStack, count : i32, color_count : i32) {
    ss.count = count;
    ss.color_count = color_count;
    for i in 0..count as usize {
        ss.slice_type[i] = fake_rand(0,color_count) as u8;
    }
}

fn ss_swapslices(ss : &mut SliceStack, i0 : usize, i1 : usize) {
    let stype : u8;
    stype = ss.slice_type[i0];
    ss.slice_type[i0] = ss.slice_type[i1];
    ss.slice_type[i1] = stype;
}

pub fn ss_flip(ss : &mut SliceStack, index : usize, direction : i32) {
    if direction > 0 {
        let end = ss.count - 1;
        let count = (ss.count - (index as i32)) / 2;
        for i in 0..count as usize {
            let mirror = (end as usize) - i;
            let pos = index + i;
            ss_swapslices(ss, pos, mirror);
        }
    } else {
        let count = (index + 1) / 2;
        for i in 0..count as usize {
            let mirror = i;
            let pos = index - i;
            ss_swapslices(ss, pos, mirror);
        }
    }
}

pub fn ss_fragmentation(ss : &SliceStack) -> i32 {
    let mut last_type : u8 = 0xFF;
    let mut frag : i32 = 0;
    for i in 0..ss.count as usize {
        if last_type != ss.slice_type[i] {
            frag += 1;
        }
        last_type = ss.slice_type[i];
    }
    return frag;
}

pub fn ss_fragmentation2(ss : &SliceStack) -> i32 {
    let mut last_index : usize = 0;
    let mut frag : i32 = 1;
    for i in 1..ss.count as usize {
        if ss.slice_type[last_index] != ss.slice_type[i] {
            frag += 1;
        }
        last_index = i;
    }
    return frag;
}

pub fn ss_iscomplete(ss : &SliceStack) -> bool {
    let mut used_type_flags : u32 = 0;
    let mut last_type : u8 = 0xFF;
    for i in 0..ss.count as usize {
        let type_flag : u32 = 1 << ss.slice_type[i];
        if (used_type_flags & type_flag) != 0
            && last_type != ss.slice_type[i] {
                return false;
        }
        used_type_flags |= type_flag;
        last_type = ss.slice_type[i];
    }
    return true;
}

pub fn ss_iscomplete2(ss : &SliceStack) -> bool {
    return (ss_fragmentation2(ss) - ss.color_count) <= 0
}

pub fn ss_find_single_joining_move(ss : &SliceStack, direction : &mut i32) -> i32 {
    let mut found_diff_type : bool;
    // left edge
    {
        found_diff_type = false;
        let outer_type = ss.slice_type[0];
        for i in 0..ss.count as usize {
            if ss.slice_type[i] == outer_type {
                if found_diff_type {
                    *direction = 0;
                    return i as i32;
                }
            } else {
                found_diff_type = true;
            }
        }
    }

    // right edge
    {
        found_diff_type = false;
        let outer_type = ss.slice_type[(ss.count - 1) as usize];
        for i in (0..(ss.count-1) as usize).rev() {
            if ss.slice_type[i] == outer_type {
                if found_diff_type {
                    *direction = 1;
                    return i as i32;
                }
            } else {
                found_diff_type = true;
            }
        }
    }
    return -1;
}

pub fn ss_find_single_joining_move2(ss : &SliceStack, direction : &mut i32) -> i32 {
    let mut last_index : usize = 0;
    // left edge
    {
        for i in 1..ss.count as usize {
            if (ss.slice_type[i] == ss.slice_type[0])
                && (ss.slice_type[i] != ss.slice_type[last_index]) {
                *direction = 0;
                return i as i32;
            }
            last_index = i;
        }
    }

    // right edge
    {
        last_index = (ss.count-1) as usize;
        for i in (0..(ss.count-2) as usize).rev() {
            if (ss.slice_type[i] == ss.slice_type[(ss.count-1) as usize])
                && (ss.slice_type[i] != ss.slice_type[last_index]) {
                *direction = 1;
                return i as i32;
            }
            last_index = i;
        }
    }
    return -1;
}

pub fn ss_find_first_double_move(ss : &SliceStack, search_dir : i32,
                                 direction : &mut i32) -> i32 {
    let mut c_index : [u8; MAX_SLICES] = [0xFF; MAX_SLICES];
    let mut last_type : u8 = 0xFF;

    if search_dir <= 0 {
        for i in 0..ss.count as usize {
            if c_index[ss.slice_type[i] as usize] == 0xFF {
                c_index[ss.slice_type[i] as usize] = i as u8;
                last_type = ss.slice_type[i];
            } else if ss.slice_type[i] != last_type {
                *direction = 1;
                return (i-1) as i32;
            }
        }
    } else {
        for i in (1..(ss.count-1) as usize).rev() {
            if c_index[ss.slice_type[i] as usize] == 0xFF {
                c_index[ss.slice_type[i] as usize] = i as u8;
                last_type = ss.slice_type[i];
            } else if ss.slice_type[i] != last_type {
                *direction = 0;
                return (i+1) as i32;
            }
        }
    }

    return -1;
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn prng_galois_lsfw_test() {
        let mut ctx : PrngCtxGaloisLsfw = PrngCtxGaloisLsfw { value:0 };
        // check seeding values
        {
            let expected_values : [u32 ; 10] = [2149580803, 1, 2149580802, 2,
                2149580801, 3, 2149580800, 4, 2149580807, 5];
            for u in 0..10 as u32 {
                prng_seed_galois_lsfw(&mut ctx, u+1);
                assert_eq!(prng_galois_lsfw(&mut ctx), expected_values[u as usize]);
            }
        }
        // check consecutive calls after initial seeding
        {
            let expected_values : [u32 ; 10] = [2149580803, 3224371202,
                1612185601, 2955673603, 3627417602, 1813708801, 3056435203,
                3677798402, 1838899201, 3069030403];
            prng_seed_galois_lsfw(&mut ctx, 1);
            for u in 0..10 as usize {
                assert_eq!(prng_galois_lsfw(&mut ctx), expected_values[u]);
            }
        }
    }

    #[test]
    fn prng_galois_lsfw_int_minmax_test() {
         let mut ctx : PrngCtxGaloisLsfw = PrngCtxGaloisLsfw { value:0x10293847 };
        // check seeding values
        {
            let expected_values : [i32 ; 10] = [ 532, 266, 133, 67, 33, 17,
                508, 754, 377, 689 ];
            for u in 0..10 as usize {
                assert_eq!(prng_galois_lsfw_int_minmax(&mut ctx, 0, 1000),
                    expected_values[u]);
            }
        }
    }

    #[test]
    fn basic_init_test() {
        let mut ss : SliceStack = SliceStack{
            count : 0,
            color_count : 0,
            slice_type : [0; MAX_SLICES]
        };
        ss_init(&mut ss, 4, 2);
        assert_eq!(4, ss.count);
        assert_eq!(2, ss.color_count);
        println!("someothing!");
    }

    #[test]
    fn ss_is_complete_test() {
        let mut ss : SliceStack = SliceStack {
            count : 4,
            color_count : 3,
            slice_type : [0; MAX_SLICES]
        };
        ss.slice_type[0] = 0;
        ss.slice_type[1] = 2;
        ss.slice_type[2] = 1;
        ss.slice_type[3] = 1;
        assert!(ss_iscomplete(&ss));
        assert!(ss_iscomplete2(&ss));
    }

    #[test]
    fn ss_fragmentation_test() {
        let mut ss : SliceStack = SliceStack {
            count : 4,
            color_count : 3,
            slice_type : [0; MAX_SLICES]
        };
        ss.slice_type[0] = 0;
        ss.slice_type[1] = 2;
        ss.slice_type[2] = 1;
        ss.slice_type[3] = 1;
        assert_eq!(3, ss_fragmentation(&ss));
        assert_eq!(3, ss_fragmentation2(&ss));
    }

    #[test]
    fn ss_find_single_joining_move_test() {
        // TODO
    }

   #[test]
    fn ss_find_first_double_move_test() {
        // TODO
    }

    #[test]
    fn lots_of_solutions() {
        const SLICE_COUNT : i32 = 16;
        const COLOR_COUNT : i32 = 8;
        let mut search_dir : i32 = 0;
        let mut direction : i32 = 0;


        const SS_COUNT : usize = 1024;
        let mut ss : [SliceStack; SS_COUNT] = [SliceStack {
            count : 0,
            color_count : 0,
            slice_type : [0; MAX_SLICES]}; SS_COUNT];

        //for s in ss.iter_mut() {
        for s in &mut ss[..] {
            ss_init(s, SLICE_COUNT, COLOR_COUNT);
        }

        for s in &mut ss[..] {
            loop {
                if ss_iscomplete(s) { break; }
                let mut index = ss_find_single_joining_move(s, &mut direction);
                if index < 0 {
                    index = ss_find_first_double_move(s, search_dir, &mut direction);
                    search_dir = !search_dir;
                }

                assert!(index != -1);
                if direction == 0 { direction = -1; }
                ss_flip(s, (index + direction) as usize, direction);
            }
        }
    }

}

/* vim: set ts=4 sts=4 sw=4 et : */

File: src/logic.c
/*
    Copyright (C) 2016  Erik Beran

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "logic.h"

#include <stdint.h> // int types, limits, macros
#include <string.h> // memset
#include <stdio.h>

#define RUN_TEST 1


//#define ARRAY_LEN(a) (sizeof(a) / sizeof(a[0]))
#define ARRAY_LEN(a) ((unsigned char*)((&a)[1]) - (unsigned char*)a)

/* start PRNG helpers */
static unsigned int l_galois_lfsr = 34;
void
seed_galois_lfsr(unsigned int s) { l_galois_lfsr = s; }
unsigned int
galois_lfsr() {
    unsigned int *const r = &l_galois_lfsr;
    *r = (*r >> 1) ^ (-(*r & 1) & 0x80200003);
    return *r;
}

// splitmix64
static uint64_t l_sm64;
void
seed_splitmix64(uint64_t s) { l_sm64 = s; }
uint64_t
splitmix64() {
    uint64_t z = (l_sm64 += UINT64_C(0x9E3779B97F4A7C15));
    z = (z ^ (z >> 30)) * UINT64_C(0xBF58476D1CE4E5B9);
    z = (z ^ (z >> 27)) * UINT64_C(0x94D049BB133111EB);
    return z ^ (z >> 31);
}

static int
fakeRand(int rmin, int rmax) {
    static float const inv_max = 1.0f / (float)UINT32_MAX;
    float const f = galois_lfsr() * inv_max;
    float const frange = (float)(rmax-rmin);
    return (f * frange + 0.5f) + rmin;
}
/* end PRNG helpers */

static void
SliceStack_SwapSlices(struct SliceStack *ss, int i0, int i1) {
    void* obj;
    int type;
    type = ss->slice_type[i0];
    obj = ss->slice_object[i0];
    ss->slice_type[i0] = ss->slice_type[i1];
    ss->slice_object[i0] = ss->slice_object[i1];
    ss->slice_type[i1] = type;
    ss->slice_object[i1] = obj;
}

//
void
SliceStack_Create(struct SliceStack *ss, int count, int color_count) {
    int i;
    //memset(ss->slice_type, 0, sizeof(SliceStack::slice_type));
    //memset(ss->slice_object, 0, sizeof(SliceStack::slice_object));
    ss->count = count;
    ss->color_count = color_count;
    for (i = 0; i < count; ++i) {
        ss->slice_type[i]  = fakeRand(0, color_count-1);
        //ss->slice_object[i] = 0;
    }
}

void
SliceStack_Randomize(struct SliceStack *ss) {
    unsigned char indices[MAX_SLICES];
    int i;

    for (i = 0; i < ss->count; i++) {
        //indices[i] = i % ss->count
    }
}

//
void
SliceStack_Flip(struct SliceStack *ss, int index, int direction) {
    int end;
    int count;
    int i;

    if (direction > 0) {
        end = ss->count - 1;
        count = (ss->count - index) / 2;
        for (i = 0; i < count; ++i) {
            int const mirror = end - i;
            int const pos = index + i;
            SliceStack_SwapSlices(ss, pos, mirror);
        }
    } else {
        //end = 0;
        count = (index + 1) / 2;
        for (i = 0; i < count; ++i) {
            int const mirror = i;
            int const pos = index - i;
            SliceStack_SwapSlices(ss, pos, mirror);
        }
    }
}

//
int
SliceStack_IsComplete(struct SliceStack *ss) {
    unsigned int used_type_flags = 0;
    unsigned int last_type = 0xFFFFFFFF;
    int i;

    for (i = 0; i < ss->count; ++i) {
        unsigned int const type_flag = 1 << ss->slice_type[i];
        if ((used_type_flags & type_flag)
            && last_type != ss->slice_type[i]) {
            return 0;
        }
        used_type_flags |= type_flag;
        last_type = ss->slice_type[i];
    }

    return 1;
}

// Fragmentation value of stack; minimum is different slice types
// i.e. if fragmentation - slice_types == 0, it's completed
int
SliceStack_Fragmentation(struct SliceStack *ss)
{
    int last_type = -1;
    int frag = 0;
    int i;

    for (i = 0; i < ss->count; ++i) {
        if (last_type != ss->slice_type[i]) {
            ++frag;
        }
        last_type = ss->slice_type[i];
    }
    return frag;
}

int
SliceStack_Fragmentation2(struct SliceStack *ss)
{
    int frag = 1;
    int i_last = 0;
    int i;

    for (i = 1; i < ss->count; ++i) {
        if (ss->slice_type[i_last] != ss->slice_type[i]) {
            ++frag;
        }
        i_last = i;
    }
    return frag;
}

//
int
SliceStack_IsComplete2(struct SliceStack *ss) {
    // Fragmentation *may* be less than color count when color count doesn't
    // represent the number of colors present in the actual stack (as apposed
    // to what it was initialized with).
    return (SliceStack_Fragmentation(ss) - ss->color_count) <= 0;
}

// Returns valid index on success
// Returns -1 on failure to find edge move
int
SliceStack_FindSingleJoiningMove(struct SliceStack *ss, int *direction) {
    int outer_type = ss->slice_type[0];
    int found_diff_type = 0;
    int i;
    //*direction = 0;

    for (i = 0; i < ss->count; ++i) {
        if (ss->slice_type[i] == outer_type) {
            if (found_diff_type) {
                *direction = 0;
                return i;
            }
        } else {
            found_diff_type = 1;
        }
    }

    // if we haven't found a matching type, try the other end and direction
    //*direction = 1;
    outer_type = ss->slice_type[ss->count - 1];
    found_diff_type = 0;
    for (i = ss->count-1; i > 0; --i) {
        if (ss->slice_type[i] == outer_type) {
            if (found_diff_type) {
                *direction = 1;
                return i;
            }
        } else {
            found_diff_type = 1;
        }
    }

    return -1;
}

int
SliceStack_FindSingleJoiningMove2(struct SliceStack *ss, int *direction) {
    int i;
    int i_last = 0;

    for (i = 1; i < ss->count; ++i) {
        if (ss->slice_type[i] == ss->slice_type[0]
            && ss->slice_type[i] != ss->slice_type[i_last]) {
            *direction = 0;
            return i;
        }
        i_last = i;
    }

    // if we haven't found a matching type, try the other end and direction
    i_last = ss->count-1;
    for (i = ss->count-2; i > 0; --i) {
        if (ss->slice_type[i] == ss->slice_type[ss->count-1]
            && ss->slice_type[i] != ss->slice_type[i_last]) {
            *direction = 1;
            return i;
        }
        i_last = i;
    }

    return -1;
}

//
int
SliceStack_FindFirstDoubleMove(struct SliceStack *ss, int search_direction,
                               int *direction) {
    unsigned char c_index[MAX_SLICES];
    int i;
    int last_type;

    memset(c_index, 0xff, sizeof(c_index));
    last_type = 0xff;

    if (search_direction <= 0) {
        for (i = 0; i < ss->count; ++i) {
            // color hasn't been recorded yet
            if (c_index[ss->slice_type[i]] == 0xff) {
                c_index[ss->slice_type[i]] = i;
                last_type = ss->slice_type[i];
                // non-contiguous color found
                // last_type will have been initialized by first iteration; don't check
                // for 0xff. If last_type == current_type we don't need to update it
            } else if (ss->slice_type[i] != last_type) {
                *direction = 1;
                return i-1;
            }
        }
    } else {
        for (i = ss->count-1; i > 0; --i) {
            // color hasn't been recorded yet
            if (c_index[ss->slice_type[i]] == 0xff) {
                c_index[ss->slice_type[i]] = i;
                last_type = ss->slice_type[i];
                // non-contiguous color found
                // last_type will have been initialized by first iteration; don't check
                // for 0xff. If last_type == current_type we don't need to update it
            } else if (ss->slice_type[i] != last_type) {
                *direction = 0;
                return i+1;
            }
        }
    }
    // This should only happen when level is complete
    return -1;
}

#if (RUN_TEST)
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
static void
RunTests() {
    int const slice_count = 16;
    int const color_count = 8;
    static int search_direction = 1;
    int direction = 0;
    int index;
    int i;

    printf("starting big init\n");
    //int const ss_count = INT32_MAX / 1024;//1024*1024;
    int const ss_count = 1024*1024;
    size_t const ss_size = sizeof(struct SliceStack) * ss_count;
    struct SliceStack* ss = (struct SliceStack*)malloc(ss_size);
    int *initial_frag = (int*)malloc(ss_count*sizeof(int));
    int *moves_to_solve = (int*)malloc(ss_count*sizeof(int));
    memset(ss, 0, ss_size);
    memset(moves_to_solve, 0, ss_count*sizeof(int));

    for (i = 0; i < ss_count; ++i) {
        SliceStack_Create(ss+i, slice_count, color_count);
        initial_frag[i] = SliceStack_Fragmentation(ss+i);
        /*int const c1 = SliceStack_IsComplete(ss+i);
        int const c2 = SliceStack_IsComplete2(ss+i);
        if (c1 != c2) { printf("%d != %d:%d\n", c1, c2, SliceStack_Fragmentation(ss+i)); }*/
    }

    printf("starting big solve\n");

    unsigned long long iters = 0;
    for (i = 0; i < ss_count; ++i) {
        direction = 0;
        while (!SliceStack_IsComplete(ss+i)) {
            ++iters;
            index = SliceStack_FindSingleJoiningMove2(ss+i, &direction);
            if (index < 0) {
                index = SliceStack_FindFirstDoubleMove(ss+i,
                                                       search_direction, &direction);
                //search_direction = !search_direction;
            }
            if (index >= 0) {
                if (direction == 0)
                    direction = -1;
                SliceStack_Flip(ss+i, index+direction, direction);
                moves_to_solve[i]++;
            }
        }
    }

    printf("iters: %" PRIu64 "\n", iters);

    printf("finished big solve\n");

    for (i = 0; i < ss_count; ++i) {
        if (initial_frag[i] * 2 < moves_to_solve[i]) {
            printf("init frag : %d moves : %d\n", initial_frag[i], moves_to_solve[i]);
        }
    }

    free(ss);
    free(moves_to_solve);
    free(initial_frag);

    printf("big solve done\n");
}

int main(int argc, char** argv) {
    RunTests();
    return 0;
}
#endif // RUN_TESTS

/* vim: set ts=4 sts=4 sw=4 et : */
